Mathlib/Algebra/CharP/LinearMaps.lean

```diff
@@ -0,0 +1,35 @@
+/-
+Copyright (c) 2024 Wanyi He. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wanyi He, Huanyu Zheng
+-/
+import Mathlib.Algebra.CharP.Subring
+
+variable {R M : Type*} [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+/-- For a commutative semiring `R` and a `R`-module `M`, if `M` contains an
+  element `x` that can eliminate `r` in `r • x = 0` (finding such element usually
+  depends on specific `•`), then the characteristic of `R` is equal to the
+  characteristic of `R`-linear endomorphism of `M`.-/
+instance char_eq_if {p : ℕ} [hchar : CharP R p]
+  (hreduction : ∃ x : M, ∀ r : R, r • x = 0 → r = 0) : CharP (M →ₗ[R] M) p where
```

Comment:
Do you really need this instance? In my opinion this `instance` is not working with instance inferring system (only works as a `theorem`), since from the conclusion `CharP (M →ₗ[R] M) p` it cannot says `hreduction` must be satisfied.

Reply:
converted to `theorem`. And do you think it is still worth keeping? The benefit of keeping it, I think, is that the injective ring homomorphism in `charP_of_injective_ringHom` might not exists for more general `R`-module `M`. So this is potentially useful somewhere else. The downside is that it kinda overlaps with `charP_of_injective_ringHom` when injective homomorphism like `Algebra.lmul` do exits, and if such function exists, `hreduction` should also be satisfied.

Reply:
converted to `theorem`. Do you think it is still worth keeping?

Reply:
To add something to @acmepjz 's comment (that is perfectly appropriate): if you type
```lean
#lint
```
at the end of your file (with `char_eq_if` still marked as `instance`) you get
```
-- Found 1 error in 4 declarations (plus 62 automatically generated ones) in the current file with 15 linters

/- The `impossibleInstance` linter reports:
SOME INSTANCES HAVE ARGUMENTS THAT ARE IMPOSSIBLE TO INFER
These are arguments that are not instance-implicit and do not appear in
another instance-implicit argument or the return type. -/
#check @char_eq_if /- argument 8 hreduction : ∃ x, ∀ (r : R), r • x = 0 → r = 0 -/
```
This error tells you that type-class inference is _uncapable_ of finding `hreduction` automatically and so the system does not even accept `char_eq_if` as an `instance`. The point here is that if there are assumptions in your declaration (like `hreduction`), for it to be an `instance` it is needed that these assumptions are found _automatically_. By this, we mean that they must themselves be part of the automatic type-search tree, while if they are simply variables the system will fail to find them. The only solution would be to create a `class hreduction` whose terms are types satisfying it, and requiring at the beginning that `R` is of this class (if makes no sense here, I am just trying to explain the whole situation). As a rule-of-thumb, you can think that (apart from type variables) all variables of a class/instance must themselves be classes.

Reply:
> converted to `theorem`. Do you think it still worth keeping?

Well, why not?

Reply:
> To add something to @acmepjz 's comment (that is perfectly appropriate): if you type
> 
> ```lean
> #lint
> ```
> 
> at the end of your file (with `char_eq_if` still marked as `instance`) you get
> 
> ```
> -- Found 1 error in 4 declarations (plus 62 automatically generated ones) in the current file with 15 linters
> 
> /- The `impossibleInstance` linter reports:
> SOME INSTANCES HAVE ARGUMENTS THAT ARE IMPOSSIBLE TO INFER
> These are arguments that are not instance-implicit and do not appear in
> another instance-implicit argument or the return type. -/
> #check @char_eq_if /- argument 8 hreduction : ∃ x, ∀ (r : R), r • x = 0 → r = 0 -/
> ```
> 
> This error tells you that type-class inference is _uncapable_ of finding `hreduction` automatically and so the system does not even accept `char_eq_if` as an `instance`. The point here is that if there are assumptions in your declaration (like `hreduction`), for it to be an `instance` it is needed that these assumptions are found _automatically_. By this, we mean that they must themselves be part of the automatic type-search tree, while if they are simply variables the system will fail to find them. The only solution would be to create a `class hreduction` whose terms are types satisfying it, and requiring at the beginning that `R` is of this class (if makes no sense here, I am just trying to explain the whole situation). As a rule-of-thumb, you can think that (apart from type variables) all variables of a class/instance must themselves be classes.

Thanks for the explanation! I noticed before in previous Github Action that there are lint mathlib errors but I didn't know what it was. Now it's clear, thanks again!

Reply:
> > converted to `theorem`. Do you think it still worth keeping?
> 
> Well, why not?

I was thinking that, it seems possible to deduce similar result via `charP_of_injective_ringHom`. For instance, we can define map `R → (M →ₗ[R] M)` as `r ↦ (fun (x : M) ↦ r • x)`. If this map is injective, then we are done. 
And its injectivity is equivalent to `∀ x : M, ∀ r : R, r • x = 0 → r = 0`, which is slightly more strict than `hreduction : ∃ x : M, ∀ r : R, r • x = 0 → r = 0`. Plus, one might also need `R` to be a ring to prove injectivity. Hence, I think that `char_eq_if` could be useful, but I'm not sure if it is significant enough to be in mathlib.

Reply:
Well, all in all I agree that this is then probably not needed, you can simply apply `charP_of_injective_ringHom`.